### PR TITLE
Rewrite convertfrom sddlstring in csharp

### PR DESCRIPTION
**Pull request by zacateras**
#3212

### CODE REVIEW

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (zacateras) [inline src/System.Management.Automation/System.Management.Automation.csproj]: this is required by types moved from System.Security.AccessControl to System.Threading.AccessControl (System.Security.AccessControl.MutexRights), am I allowed to add new references here ?

Comment (daxian-dbw) [inline src/System.Management.Automation/System.Management.Automation.csproj]: It looks to me the package `System.Threading.AccessControl` is required by `Microsoft.PowerShell.Commands.Utility` instead of `System.Management.Automation`. If that's the case, this `PackageReference` entry should be added to `Microosft.PowerShell.Commands.Utility` instead. And please remove the same entry from `Microsoft.PowerShell.SDK.csproj`.

**Comment by daxian-dbw** [comment]
@iSazonov if you have time, could you please review this PR?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: Oh, we re-evaluate Enum.GetNames and Enum.Parse again and again. We should refactor "GetAccessRights" to use cache(s).

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: We should use StringBuilder because `aceString` is changed multiple times.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: Please add Newline.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs]: It seems we can directly use `UtilityCommonStrings.TypeNotSupported`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @SteveL-MSFT @daxian-dbw We should make a conclusion about the algorithm. SDDL is always "object-specific" but we go through all "types". I believe that's a mistake. An user always know the object type (where he did get the SDDL), should set it explicit and the cmdlet must decode only against this type.

**Review by iSazonov — COMMENTED** [review]
Leave a comment

**Comment by iSazonov** [comment]
@zacateras Thanks for your work! I have question about the algorithm that I addressed to PowerShell team - please wait the conclusion before continue.

Also we haven't tests for the cmdlet - you should add its and pass CI before we can merge.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: We could write a script to generate C# and avoid reflection entirely.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: I don't understand this domain enough to be confident here - but in principle I fully agree with @iSazonov - the current algorithm is strange.

**Review by lzybkr — COMMENTED** [review]
Thanks for the contribution - I'll be very happy to see this one merged.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @lzybkr Could you please ping an expert in this domain?

File: src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1
Comment (zacateras) [inline src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1]: this file name is referenced 3 times:
* /src/Modules/map.json (remove entry?)
* /src/System.Management.Automation/engine/InitialSessionState.cs (replace with some other build result file?)
* /src/System.Management.Automation/security/Authenticode.cs (replace with some other build result file?)

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: Following up on my own question and after playing around with this cmdlet a little, we definitely want strongly typed output.

File: src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1
Comment (lzybkr) [inline src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1]: Thanks for catching that.
Remove the entry in `map.json` - that's perfectly safe.
I think you can change the other two references to `Microsoft.PowerShell.Utility.psd1`.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: The [documentation for SDDL](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379567(v=vs.85).aspx) looks pretty good, I think if we have concerns about this algorithm, we can answer them ourselves, but I don't have time for that now.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Can you add some tests?  I just cloned this branch and a basic attempt to use this cmdlet failed:

```powershell
PS C:\users\slee\repos\zacateras> $sddl = "O:BAG:BAD:(A;CI;CCDCLCSWRPWPRCWD;;;BA)(A;CI;CCDCRP;;;NS)(A;CI;CCDCRP;;;LS)(A;
CI;CCDCRP;;;AU)"
PS C:\users\slee\repos\zacateras> ConvertFrom-SddlString -Sddl $sddl
ConvertFrom-SddlString : Unable to cast object of type 'System.Security.AccessControl.FileSystemRights' to type
'System.Int64'.
At line:1 char:1
+ ConvertFrom-SddlString -Sddl $sddl
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-SddlString], InvalidCastException
    + FullyQualifiedErrorId : System.InvalidCastException,Microsoft.PowerShell.Commands.ConvertFromSddlStringCommand
```

Which worked with the old script cmdlet:

```powershell
PS C:\users\slee\repos\PowerShell> $sddl = "O:BAG:BAD:(A;CI;CCDCLCSWRPWPRCWD;;;BA)(A;CI;CCDCRP;;;NS)(A;CI;CCDCRP;;;LS)(A
;CI;CCDCRP;;;AU)"
PS C:\users\slee\repos\PowerShell> ConvertFrom-SddlString -Sddl $sddl


Owner            : BUILTIN\Administrators
Group            : BUILTIN\Administrators
DiscretionaryAcl : {NT AUTHORITY\Authenticated Users: AccessAllowed (ListDirectory, WriteData,
                   WriteExtendedAttributes), NT AUTHORITY\LOCAL SERVICE: AccessAllowed (ListDirectory, WriteData,
                   WriteExtendedAttributes), NT AUTHORITY\NETWORK SERVICE: AccessAllowed (ListDirectory, WriteData,
                   WriteExtendedAttributes), BUILTIN\Administrators: AccessAllowed (ChangePermissions,
                   CreateDirectories, ExecuteKey, ListDirectory, ReadExtendedAttributes, ReadPermissions, Traverse,
                   WriteData, WriteExtendedAttributes, WriteKey)}
SystemAcl        : {}
RawDescriptor    : System.Security.AccessControl.CommonSecurityDescriptor
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @lzybkr what kind of code generation script do you mean? t4 template?

Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: No, I just meant a PowerShell script that generates switch statements or something like that.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: I re-read the document and made any tests - My conclusion is full discard the algorithm:

1. From the [docs](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379557%28v=vs.85%29.aspx):
> Each type of securable object defines its own set of specific access rights and its own mapping of generic access rights. For information about the specific and generic access rights for each type of securable object, see the overview for that type of object.

- the conclusion is  an SDDL has no meaning without object entity - the cmdlet should get the object type as argument.

2. The cmdlet calls  [Security.AccessControl.CommonSecurityDescriptor]::new($false,$false,$Sddl)
[The constructor](https://github.com/dotnet/corefx/blob/0e4095bc5da407df6cbe0aa350e1b2ccf32696a4/src/System.Security.AccessControl/src/System/Security/AccessControl/SecurityDescriptor.cs#L949) in Core FX

I tested the cmdlet with (1)  [Security.AccessControl.CommonSecurityDescriptor]::new($false,$false,$Sddl) and (2)  [Security.AccessControl.CommonSecurityDescriptor]::new($true,$false,$Sddl) for path "c:\test" (it is container) and got different results.

- the conclusion is the cmdlet should take into account `isContainer` (folders/containers) and `isDS` (directory objects and containers).

3. CoreFX can throw exceptions (ArgumentException) - maybe the cmdlet should catch them and re-throw (put CoreFX exception in InnerException)

4.  The cmdlet should have strongly typed output. In current PSCustomObject we should add "Type".

5. "Type" - we can define it as Enum ( remove unsupported member by #if !Unix ) and exclude ValidateSet at all.

6. We should add tests for all object types.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: used custom enum instead

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: thanks, done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: added static cache field

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: added ps1 generator

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: used StringComparer.OrdinalIgnoreCase

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs]: removed obsolete property

File: src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1
Comment (zacateras) [inline src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1]: done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: removed ValidateSet, used custom enum instead

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: ok

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @iSazonov I also have doubts about current method of converting ACE AccessMask into access flag enum based strings. There are situations when flag cannot be converted without additional logic i.e. [FileSystemRights](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs#L83&L85) contains entries with duplicate flag values - some applies only for files, other only for containers. I guess that this won't be solved by requiring Type parameter.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @zacateras Please see (2) point from my conclusion.
> •the conclusion is the cmdlet should take into account  isContainer  (folders/containers) and  isDS  (directory objects and containers).

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @lzybkr Will this require recompilation if there will are changes in CoreFX? We could initilize this in static cache on Begin step.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by iSazonov** [comment]
@SteveL-MSFT The algorithm is broken! Please see my comment above https://github.com/PowerShell/PowerShell/pull/3936#discussion_r121588232 I believe we cannot merge the PR.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Reviewing algorithm

**Comment by iSazonov** [comment]
@daxian-dbw Can we use [System.DirectoryService](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.DirectoryServices) or before we should move to 4.5.0 version ?

**Comment by iSazonov** [comment]
@zacateras Do you plan to continue?

**Comment by zacateras** [comment]
@iSazonov yep. I thought that we are waiting for some external review.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs
Comment (zacateras) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs]: @iSazonov Apart from fixing CommonSecurityDescriptor constructor call (by requiring isContainer and isDS flags) we should also take care about correct parsing of AccessMasks to enum names. FileSystemRights and ActiveDirectoryRights contain several names that use the same value, eg. FileSystemRights contain AppendData & CreateDirectories which have [the same value (4)](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs#L83&L85). In consequence Directories with access flag (4) can get AppendData right, what is incorrect.

As a solution we can add custom enums (or at least custom mapping dictionaries) containing distinct subsets of allowed values - one for containers (ContainerSystemRights) and one for other objects (FileSystemRights). Alternatively we can discard CommonSecurityDescriptor/Right-Enum based parsing algorithm entirely and try to implement text parser using sddl docs. What do you think ?

**Comment by SteveL-MSFT** [comment]
Now that Beta.5 is out, I'll spend time on this tomorrow

**Comment by iSazonov** [comment]
We should follow MSDN docs - see my comment above.

**Comment by iSazonov** [comment]
@SteveL-MSFT Could you please continue?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.Generate.ps1
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.Generate.ps1]: Was this file meant to be checked in?  It's not part of the build to run and the resulting source doesn't indicate it shouldn't be edited.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.cs
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.cs]: since we're officially removing fullclr support, we should not include this

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.cs
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromSddlStringCommand.RightType.cs]: since we're officially removing fullclr support, we should not include this

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-SddlString.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-SddlString.Tests.ps1]: Since this cmdlet is not supported on non-Windows, you should skip them in bulk using this pattern https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md#skipping-tests-in-bulk (that way if additional tests get added they get auto-skipped)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-SddlString.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-SddlString.Tests.ps1]: Should be some tests for `-Type` parameter

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: This doesn't appear to be a supported package, perhaps we should not support AD rights for now.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
Sorry this took so long, got randomized by other things.  I also compared the result with 

```powershell
ConvertFrom-SddlString "D:(A;;FA;;;SY)(A;;FA;;;BA)"
```

```
# old script version
DiscretionaryAcl : {NT AUTHORITY\SYSTEM: AccessAllowed (ChangePermissions, CreateDirectories, Delete, DeleteSubdirectoriesAndFiles, ExecuteKey, FullControl,
                   FullControl, FullControl, FullControl, ListDirectory, Modify, Read, ReadAndExecute, ReadAttributes, ReadExtendedAttributes,
                   ReadPermissions, Synchronize, TakeOwnership, Traverse, Write, WriteAttributes, WriteData, WriteExtendedAttributes, WriteKey),
                   BUILTIN\Administrators: AccessAllowed (ChangePermissions, CreateDirectories, Delete, DeleteSubdirectoriesAndFiles, ExecuteKey, FullControl,
                   FullControl, FullControl, FullControl, ListDirectory, Modify, Read, ReadAndExecute, ReadAttributes, ReadExtendedAttributes,
                   ReadPermissions, Synchronize, TakeOwnership, Traverse, Write, WriteAttributes, WriteData, WriteExtendedAttributes, WriteKey)}
```

```powershell
# c# version
DiscretionaryAcl : {NT AUTHORITY\SYSTEM: AccessAllowed (AppendData, ChangePermissions, CreateFiles, Delete, DeleteSubdirectoriesAndFiles, ExecuteFile,
                   FullControl, FullControl, FullControl, FullControl, Modify, Read, ReadAndExecute, ReadAttributes, ReadData, ReadExtendedAttributes,
                   ReadKey, ReadPermissions, Synchronize, TakeOwnership, Write, WriteAttributes, WriteExtendedAttributes, WriteKey), BUILTIN\Administrators:
                   AccessAllowed (AppendData, ChangePermissions, CreateFiles, Delete, DeleteSubdirectoriesAndFiles, ExecuteFile, FullControl, FullControl,
                   FullControl, FullControl, Modify, Read, ReadAndExecute, ReadAttributes, ReadData, ReadExtendedAttributes, ReadKey, ReadPermissions,
                   Synchronize, TakeOwnership, Write, WriteAttributes, WriteExtendedAttributes, WriteKey)}
```

And the newer version has more permissions (although not sure if the old one was wrong).

**Comment by iSazonov** [comment]
@SteveL-MSFT Did you look [my comment](https://github.com/PowerShell/PowerShell/pull/3936#discussion_r121588232) above? Main is "SDDL has no meaning without object entity - the cmdlet should get the object type as argument."

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: Official package is  [System.DirectoryService](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.DirectoryServices). It is 4.5.0 version. @daxian-dbw said that we should review the API before use.

**Comment by SteveL-MSFT** [comment]
@iSazonov Seems like if you have the object, you would be using `Get-Acl`.  `ConvertFrom-SDDLString` is more likely used when you have just the SDDL and just want to understand (perhaps roughly) what it represents.  Not sure how useful this is if you require the object.

**Comment by iSazonov** [comment]
@SteveL-MSFT My idea is that we should make `-Type` mandatory.

**Comment by zacateras** [comment]
@iSazonov If AccessMask is presented using [well known aliases](https://github.com/tpn/winsdk-10/blob/master/Include/10.0.10240.0/shared/sddl.h#L126) then it can be parsed entirely without the knowledge of securable object type (Type parameter).

In case of numerical AccessMasks (eg. 0x02142512) we can translate only higher 16-32bits [generic part](https://msdn.microsoft.com/pl-pl/library/windows/desktop/aa374896(v=vs.85).aspx) if the Type is not provided (lower bits can remain unparsed as a reminder). If the Type parameter is present then specific rights scheme could be used - according to constant definitions in [winnt.h](https://source.winehq.org/source/include/winnt.h). Also more securable object types can be considered - according to the list in [docs](https://msdn.microsoft.com/en-us/library/aa379557(VS.85).aspx).

I started an implementation of SDDL parser using the documentation. The code can be found [here](https://github.com/zacateras/sddl-parser) (with examples [here](https://github.com/zacateras/sddl-parser/blob/master/test/Sddl.Parser.Tests/SddlTests.cs)). If we want to use parts of the implemented approach then I can integrate its simplified form into the command.

What do you think about such solution?

**Comment by iSazonov** [comment]
@zacateras I still don't understand the need for this heuristic. For which scenarios is it addressed? As for me it looks like a trasformation enum1 -> int -> enum2 - what's confusing.

**Comment by daxian-dbw** [comment]
This PR has been left unattended for a long time. I will close this PR for now.
@zacateras Please feel free to reopen it if you decide to continue on it at a later time.

**Pull request closed without merging**
